### ISSUE TITLE
redundant_closure_call false positive when closure is returned from a macro

### ISSUE THREAD

**Issue by halvko** [rust-lang/rust-clippy#12358]
### Summary

If using a macro which expands to a closure, instantly calling the closure produced by that macro triggers the lint

### Lint Name

redundant_closure_call

### Reproducer

I tried this code:

```rust
macro_rules! get_env {
    ($env_name:expr) => {
        || {
            env::var($env_name)
                .map_err(|e| log::warn!("environment variable {}: {e}", $env_name))
                .ok()
        }
    };
}

static NAME: once_cell::sync::Lazy<Option<u64>> =
    Lazy::new(|| get_env!("NAME")().flat_map(|s| s.parse::<u64>().map_err(|e| log::warn("could not parse NAME: {e}")).ok()));
```

I saw this happen:

```
warning: try not to call a closure in the expression where it is declared
  --> src/lib.rs:16:5
   |
16 |     get_env!("NAME")().and_then(|s| {
   |     ^^^^^^^^^^^^^^^^^^ help: try doing something like: `get_env!("NAME")`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_closure_call
   = note: `#[warn(clippy::redundant_closure_call)]` on by default
```

I expected to see this happen:

Nothing

### Version

```text
rustc 1.76.0 (07dca489a 2024-02-04)
binary: rustc
commit-hash: 07dca489ac2d933c78d3c5158e3f43beefeb02ce
commit-date: 2024-02-04
host: x86_64-unknown-linux-gnu
release: 1.76.0
LLVM version: 17.0.6
```

### Additional Labels

@rustbot label +I-suggestion-causes-error
